pkg/util/workloadrepo/utils.go

```diff
@@ -16,47 +16,62 @@ package workloadrepo
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"strconv"
 	"strings"
 	"time"
 
 	"github.com/pingcap/tidb/pkg/meta/model"
-	"github.com/pingcap/tidb/pkg/util/slice"
 )
 
-func generatePartitionDef(sb *strings.Builder, col string) {
+func generatePartitionDef(sb *strings.Builder, col string, now time.Time) {
 	fmt.Fprintf(sb, " PARTITION BY RANGE( TO_DAYS(%s) ) (", col)
 	// tbInfo is nil, retval must be false
-	_ = generatePartitionRanges(sb, nil)
+	_, _ = generatePartitionRanges(sb, nil, now)
 	fmt.Fprintf(sb, ")")
 }
 
-func generatePartitionRanges(sb *strings.Builder, tbInfo *model.TableInfo) bool {
-	now := time.Now()
-	newPtNum := 2
-	// add new partitions per day
-	// if all partitions to be added existed, do nothing
-	allExisted := true
-	for i := range newPtNum {
-		// TODO: should we make this UTC? timezone issues
-		newPtTime := now.AddDate(0, 0, i+1)
-		newPtName := "p" + newPtTime.Format("20060102")
-		if tbInfo != nil {
-			ptInfos := tbInfo.GetPartitionInfo().Definitions
-			if slice.AnyOf(ptInfos, func(i int) bool {
-				return ptInfos[i].Name.L == newPtName
-			}) {
-				continue
+func generatePartitionName(t time.Time) string {
+	return "p" + t.Format("20060102")
+}
+
+func parsePartitionName(part string) (time.Time, error) {
+	return time.ParseInLocation("p20060102", part, time.Local)
+}
+
+func generatePartitionRanges(sb *strings.Builder, tbInfo *model.TableInfo, now time.Time) (bool, error) {
+	firstPart := time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, time.Local)
+	if tbInfo != nil {
+		pi := tbInfo.GetPartitionInfo()
+		if pi != nil {
+			ptInfos := pi.Definitions
+			lastPartDate, err := parsePartitionName(ptInfos[len(ptInfos)-1].Name.L)
+			if err != nil {
+				// not sure what to do
+				return true, err
+			}
+			if firstPart.Before(lastPartDate) {
+				firstPart = lastPartDate
```

Comment:
I'm a little confused by what "first" and "last" mean here. The name `firstPart` implies the oldest partition, but as written, `firstPart` seems to be used more like I would expect `lastPart` (the newest partition) to be.

Reply:
I simplified the code a bit.  It was a mess before.